### ISSUE TITLE
RFC: structure of SciPy 2.0 paper

### ISSUE THREAD

**Issue by j-bowhay** [scipy/scipy-articles#275]
To get discussion started, here is a proposal for the structure of the paper. I'm very open to changing it—this is mainly intended as a catalyst for discussion.

The headings below represent the proposed section structure. The section titles are placeholders and the content is just rough notes on what the section might include.

**Aim / Scope of paper:** To motivate and describe the advances that have been made to SciPy since version 1.0.

- [ ] **Introduction**: Brief background on SciPy, its aims and scope. Refer to the 1.0 paper for the project's history. Motivate the 2.0 release.
- [ ] **Support for GPUs and just-in-Time compilation through the Python array API standard**: Motivation, key ideas, coverage, performance improvements
- [x] **Translation of Fortran codes into other languages**: xref #278 Focus on history and problems we aimed to solve rather than Fortran bashing (along the lines of https://github.com/scipy/scipy-articles/issues/278#issuecomment-5208511567)
- [ ] **Adaptive finite difference numerical differentiation of black-box functions** (`scipy.differentiate`)
- [ ]  **Improved integral transforms** (`scipy.fft`): Explain why new FFT interface was needed and benefits.
- [ ] **Improved infrastructure for statistical distributions**: Explain history of old infra and its short comings. Explain new functionality
- [ ]  **Quasi-Monte Carlo sampling and sensitivity analysis**: Explain motivation and new functionality
- [ ] **Random number generation from statistical distributions**: Explain motivation and new functionality
- [ ] **An updated interface for sparse arrays**

Explain history and need to move to new sparse array interface.

# Other improvements and new functionality

## `scipy.integrate`

New functionality: `cubature`, `lebedev_rule`, `cumulative_simpson`, `qmc_quad`, `quad_vec`

## `scipy.interpolate`

New functionality: `AAA`, `FloaterHormannInterpolator`, `make_splrep`, `make_splprep`, `generate_knots`, `make_smoothing_spline`, Modified Akima interpolation, `NdBSpline`

## `scipy.linalg`

ILP64 support

Accelerate support

Batched input support

structure detection for solve etc

New functionality: bandwidth, issymmetric, ishermitian, matmul_toeplitz, cossin, khatri_rao, convolution_matrix, fiedler / fiedler_companion, ldl, null_space, cdf2rdf, rsf2csf

## `scipy.ndimage`

New functionality: vectorized_filter, value_indices - Maybe this isn't enough to warrent mentioning?

## scipy.optimize

HiGHS

Migration of COBYLA to PRIMA
New global optimisers: direct, dual_annealing, shgo

Other new functionality: toms748, trust-constr, cobyqa, isotonic_regression, root_scalar

Improved paralellisation

## `scipy.signal`

ShortTimeFFT / closest_STFT_dual_window

New windows:lanczos, kaiser_bessel_derived, taylor, dpss, general_cosine and general_hamming

Peak finding: find_peaks / peak_prominences / peak_widths

Other new functionality: Whittaker-Henderson smoothing, firwin_2d, envelope, CZT, ZoomFFT, gammatone, iircomb, correlation_lags, oaconvolve

ZPK versions of frequency transformations: scipy.signal.bilinear_zpk, scipy.signal.lp2bp_zpk, scipy.signal.lp2bs_zpk, scipy.signal.lp2hp_zpk, scipy.signal.lp2lp_zpk (maybe)

## `scipy.sparse`

New functionality: nD coo arrays

New sparse linalg functionality: n-dimensional linear operators added to scipy.sparse.linalg.LinearOperator, funm_multiply_krylov, PROPACK, tfqmr

New sparse `csgraph` functionality: `yen`, `min_weight_full_bipartite_matching`

## `scipy.spatial`

`transform.RigidTransform` and many other improvements to `spatial.transform`

`geometric_slerp`

## `scipy.special`

`xsf`

Improved accuracy of many functions by wrapping Boost

Improved Legendre function implementations with a more consistent interface

New functionality: log_wright_bessel, stirling2, betaincc, betainccinv, powm1, log_expit, ellipr{c,d,f,g,j}, wright_bessel, ndtri_exp, log_softmax, voigt_profile, softmax, owens_t.

## scipy.stats

New functionality: tons!! I will let the stats maintainers organise this...

## Elementwise-iterative methods

Explain new functionality added and its benefit/usecase over normal solvers.

## free-threaded Python support

Is this significant enough?

## Improved developer tooling / security / infra?
(I don't know if these are too developer focused?)

separate scipy-release repo (security)
github actions (move off appveyor/travis)
trusted publishing to PyPI (security)
new platforms: manylinux-aarch64, win-arm64, macos-arm64, musllinux-x86_64, etc. We've come up with rationalisation for tiered support.
Developed AI policy.

# Concluesion

Conclude with what are the next priorities for SciPy

**To reiterate, this is very much a first draft and I would be very happy suggestions / alternative proposals!**

**Comment by chrisb83** [comment]
I am adding a summary of the structure of the 1.0 article to get some further inspiration

- Introduction (what is SciPy? Release of version 1.0, how SciPy evolved to a mature library)
- Background (about Python, how it became popular for numerical computations, followed by the evolution of SciPy in the three following subsections)
  -  SciPy begins
  - ScipPy matures
  - SciPy today (including stats about lines of code, number of packages that depend on SciPy et)
- Architecture and implementation choices
  - Project scope (what kind of algorithms, SciPy vs other packages like statsmodels, sklearn)
  - Package organization (brief overview of all subpackages)
  - Language choices (stats on languages used, preference Cython vs C vs C++ vs Fortran)
  - API and ABI evolution
- Key technical improvements (summary of improvements over the past three years, grouped by different topics like "Data structures", ''Numerical optimization", "Test and benchmark suite")
- Project organization and governance structure
  - Governance (Governance Document, Steering Council, BDFL, Code of Conduct)
  - Maintainers and contributors (how many contributors per release, how many maintainers, who can become a maintainer/contributor, mentions development guide)
  - Funding
  - Downstream projects
- Discussion (stats on the popularity of SciPy, examples of prominent projects using SciPy, outlook/roadmap)

**Comment by chrisb83** [comment]
Potentially the following sections are worth keeping with updated content:

- Architecture and implementation choices: in that context, the removal of Fortran can be explained
- Project organization and governance structure

**Comment by j-bowhay** [comment]
Here is the description from ACM TOMS:

> **Software Release paper**
>
> A paper that describes major developments, either new or updated features, in the implementation of a widely used software package that is relevant to the TOMS community.
>
> The developments described in these papers need not be new theoretical advances; rather, they should bring significant additional capabilities to an existing user base.
>
> Papers in this category provide authors of these packages a venue to discuss what is new in a holistic way; thus, they inform users of these packages about ongoing development and provide them with a target to cite something specific to one or a small number of package releases. They also allow developers who contributed to a recent release to see their work on the software acknowledged by being co-authors on a release paper.
>
> Release papers are typically anchored in a previously published paper that describes the overall design of a package, are short (typically fewer than 10 pages), and are reviewed on a separate, expedited track.
>
> Review criteria include the presentation of the material with an eye towards software package users, and whether the package and the release(s) described in the manuscript are substantial contributions to mathematical software; reviewers may assess the quality of the published software. Having an open-source implementation is considered an advantage.

So while I do think we should comment on both those things, I think the main thrust of the paper should be based around new capabilities.

**Comment by j-bowhay** [comment]
I updated the section headings with slightly more descriptive titles

**Comment by ilayn** [comment]
We might want to downplay the Fortran thing since there is a dedicated fandom related to it and they don't really read much if there is criticism involved but immediately start typing how good modern fortran isand we are ignorant etc. It might create some unnecessary noise.

I mean I don't mind shouting from the rooftop and I love wrestling in the mud so don't get me wrong. But not all authors might like the controversial stuff.

**Comment by lucascolley** [comment]
yeah, I think taglines like 'fortran free' are to be avoided when not used in combination with explanation of the kind of fortran from which we are free.

**Comment by j-bowhay** [comment]
> We might want to downplay the Fortran thing since there is a dedicated fandom related to it and they don't really read much if there is criticism involved but immediately start typing how good modern fortran isand we are ignorant etc. It might create some unnecessary noise.
> 
> I mean I don't mind shouting from the rooftop and I love wrestling in the mud so don't get me wrong. But not all authors might like the controversial stuff.

That is a good point... you can just imagine being assigned a reviewer 2 who is a die hard Fortran fan...

~~I'll remove having it as such a prominent section but maybe we can slip a sentence in somewhere.~~ It seems a shame for a herculean effort to go unmentioned, especially now we are reaping the rewards of it with things like `xsf`.

**Comment by ev-br** [comment]
FWIW the Fortran subject made a bit of a splash in my SciPy 2.0 talk at EuroSciPy a couple weeks ago. I think we definitely should discuss it in some detail.

**Comment by mdhaber** [comment]
I wonder if we can/should group together some things under the banner of "vectorization" instead of al having a dedicated "elementwise iterative methods" section.  Batched linear algebra, ndimage `vectorized_filter`, and many hypothesis test improvements could also go there, for example.

And as much as I like to see stats stuff highlighted, I think we could group some things together under a heading of random sampling. Maybe more if we have a good name that describes fundamental probability distribution tools - sampling, but also fitting and evaluating functions of probability distributions.

Maybe if there were that and a separate section for frquentist inference, the `stats` section itself could be rather small, mentioning stuff like improvements in working with missing data - or we could go without a `stats` section. I'm not sure we can hope to be comprehensive, so maybe we can only mention the things that are particularly important on their own or can be grouped with others.

**Comment by DietBru** [comment]
Not sure if this is in the scope of this article: Since we make up one of the larger scientific Python projects, readers might be also interested in the people side of SciPy. Specifically:

* Are there already lessons learned from dealing with the AI-generated contributions?
* The rate of PRs is slowly but steadily growing. Is that healthy for SciPy and its community?
* Were there noteworthy events occurring between SciPy 1.0 and 2.0?

For procrastination purposes, I built a dataset of SciPy-PRs: The following plot shows the number of merged PRs per month (note that the last month 2026/08 is incomplete):

<img width="1027" height="354" alt="Image" src="https://github.com/user-attachments/assets/366f7536-950e-433b-8839-976c001c7a83" />

Compare this to the monthly closed unmerged PRs:

<img width="1027" height="354" alt="Image" src="https://github.com/user-attachments/assets/7f1fb8b9-7a8b-46ad-a4b2-f65c6101589b" />

The following plot shows the monthly rate of contributors' first PRs:

<img width="1027" height="354" alt="Image" src="https://github.com/user-attachments/assets/faaafd6a-f638-4bc1-ba61-545dd143c7c0" />



The increases of new contributors and closed PRs in 2026 are surely a symptom of AI-based contributions. I would have expected that the increase would be a little bit more dramatic. The [marimo] worksheet used to generate this plot as well as the PR-dataset can be found [here]. Are there any ideas about what other useful metrics which could be extracted from that PR-dataset?  

[marimo]: https://marimo.org/
[here]: https://github.com/DietBru/PR-Statistics_SciPy

**Comment by chrisb83** [comment]
> I wonder if we can/should group together some things under the banner of "vectorization" instead of al having a dedicated "elementwise iterative methods" section. Batched linear algebra, ndimage `vectorized_filter`, and many hypothesis test improvements could also go there, for example.
> 
> And as much as I like to see stats stuff highlighted, I think we could group some things together under a heading of random sampling. Maybe more if we have a good name that describes fundamental probability distribution tools - sampling, but also fitting and evaluating functions of probability distributions.
> 
> Maybe if there were that and a separate section for frquentist inference, the `stats` section itself could be rather small, mentioning stuff like improvements in working with missing data - or we could go without a `stats` section. I'm not sure we can hope to be comprehensive, so maybe we can only mention the things that are particularly important on their own or can be grouped with others.

I created a new issue with an overview of changes to the stats module: https://github.com/scipy/scipy-articles/issues/279

**Comment by lucascolley** [comment]
> I think the main thrust of the paper should be based around new capabilities

With this in mind, I think DX work like that on the Pixi workspace would be off-topic. It would be nice, though, if we could find a way to mention the deprecation work spearheaded by @h-vetinari and @j-bowhay on https://github.com/scipy/scipy/issues/15765, and my work on sane vendoring practices at https://github.com/scipy/scipy/tree/main/subprojects — there has been a lot of work in these areas which while not on new features has helped get rid of some of the ugliest parts of the repo since 1.0.

**Comment by dschmitz89** [comment]
Wow, I was not aware that you were already so far with the paper preparation @j-bowhay , thanks. Are you planning a section on general improvements by module? I could take a stab at that for `optimize`. I was not really part of the array API and Fortran removal movements, so cannot really help out there.

**Comment by j-bowhay** [comment]
> Are you planning a section on general improvements by module?

Listing improvements by module is something I've actually been trying to avoid. I think if we did this the paper would just read like a giant aggregated release note. Instead, I think the sections should describe a group of feature/improvements/changes under a common theme. For example, QMC covers `stats.qmc` but also `integrate.qmc_quad` and other related things. I am aiming for a 10-20 page exposition of the most impactful changes for users. This is very unlikely to be completely exhaustive in covering every improvement.

> I could take a stab at that for `optimize`.

Is there a common theme or group of new features in `optimize` that could constitute a section and are of high impact? (I haven't checked yet to confirm either way)

> I was not really part of the array API and Fortran removal movements, so cannot really help out there.

General proof reading and review for other sections is very welcome regardless!

**Comment by dschmitz89** [comment]
> > Are you planning a section on general improvements by module?
> 
> Listing improvements by module is something I've actually been trying to avoid. I think if we did this the paper would just read like a giant aggregated release note. Instead, I think the sections should describe a group of feature/improvements/changes under a common theme. For example, QMC covers `stats.qmc` but also `integrate.qmc_quad` and other related things. I am aiming for a 10-20 page exposition of the most impactful changes for users. This is very unlikely to be completely exhaustive in covering every improvement.
> 
> > I could take a stab at that for `optimize`.
> 
> Is there a common theme or group of new features in `optimize` that could constitute a section and are of high impact? (I haven't checked yet to confirm either way)
> 
> > I was not really part of the array API and Fortran removal movements, so cannot really help out there.
> 
> General proof reading and review for other sections is very welcome regardless!

Well, the improvements in `optimize` are great but not earth shattering, so let's focus on the points you are already looking into.

I will gladly help with reviews.
